### DISCUSSION TITLE
Improved inline script support

### DISCUSSION THREAD

**Discussion by delucis** [withastro/roadmap#995] [Proposal]
# Summary

Adding support for minification and TypeScript to `<script is:inline>` would improve authoring experience and output performance.

# Background & Motivation

Astro has an `is:inline` directive for script tags that prevents a tag from being bundled and hoisted. Currently, the contents of an inline script are not processed at all and shipped verbatim to the browser.

One common use case for inline scripts is to ensure some code executes at a specific point in the page load lifecycle. The classic example is colour scheme preference, where an inline script should execute and set a dark/light mode preference as early as possible to avoid a flash of the incorrect colour scheme.

Astro’s current behaviour means you are forced to write your inline script in plain JavaScript and (ideally) author it in a fairly terse way, because all comments etc. will be shipped to users as is. This makes for a less good authoring/maintenance experience, but also for a worse end-user experience if they have to download larger than necessary script tags.

As an example, Starlight has multiple places where it uses inline scripts to control page behaviour during the page load lifecycle and each of these needs to be authored carefully and in a less maintainable way to avoid shipping overly verbose inline code.

# Goals

- To reduce the size of JavaScript shipped to end users even when it is inlined.
- To support better authoring practices in inline scripts.

# Example

The proposal here would be to process inline scripts without bundling and hoisting them. Ideally this would include support for TypeScript types as well.

For example, a user might author:

```astro
<script is:inline>
/** Multiply a number by five. */
const timesFive = (variable: number) => {
  const constant = 5;
  return variable * constant;
};
</script>
```

And the build output could look something like:

```html
<script>const timesFive=(a)=>a*5</script>
```

This would enable people to author with inline comments, expressive variable naming, types, etc., while still shipping a minimal script with the same functionality to the browser. (This would need to use a minification pipeline that respects the behaviours of global HTML scripts, e.g. in the example above the `timesFive` identifier will be attached to `globalThis`, so should not be minified in case another script somewhere is relying on it.)

Other restrictions of inline scripts would still apply. For example an `import` statement would not be bundled.

**Comment by matthewp** [comment]
What happens if someone does:

```astro
<script is:inline src="https://somecdn.com/myscript.js"></script>
```

**Reply by delucis** [reply]
Nothing I think? There’s probably a similar question around:

```astro
<script is:inline set:html={`console.log(${serverVariable})`}></script>
```

**Reply by delucis** [reply]
One option could also be to basically split behaviour into three tiers:

- `<script>` — bundled, emitted, as currently
- `<script is:inline>` — processed, but injected inline
- `<script is:raw>` — for when you really want something Astro doesn’t touch (works today already)

**Reply by spaceemotion** [reply]
I think having the `set:html` bits still going through bundling steps would be helpful for cleaning up any generated code (e.g. minifying the output).

**Reply by delucis** [reply]
The main issue is the difference between these two:

```astro
<script is:inline>console.log('example')</script>
```
```astro
<script is:inline set:html={`console.log('example')`}></script>
```

Superficially they’re the same, but because `set:html` supports any JS value they’re not in practice. Astro does script processing at build time, but the dynamic JS value passed to `set:html` would need to be processed at render time, because people could do something like this, where the final value of the prop is only available when the component is rendered:

```astro
---
const dynamicValue = Astro.url.searchParams.get('example');
---
<script is:inline set:html={`console.log(${JSON.stringify(dynamicValue)})`}></script>
```

That rules out build-time processing of the `set:html` value unfortunately.

**Comment by sommerper** [comment]
Completely agree. The worst part is that setting `define:vars` makes it inline an invalidates all the types that I've written and I have to rewrite my code.

Basically I can't ever be 100% sure that I can use typescript if I suddenly find myself in need of a server defined var.

**Comment by delucis** [comment]
Adding another motivation for us to think about: CSP support. Currently, inline scripts are not handled by Astro's CSP feature. Moving these to be processed might give us the opportunity to automatically add hashes for people rather than forcing them to move to a very manual approach if they want to enable CSP.

**Reply by Adammatthiesen** [reply]
Expanding on this idea, plus my own wants/desires... it would be nice if we could have a proper way to pass vars to in-lined scripts... https://github.com/withastro/roadmap/discussions/995#discussioncomment-10337523 I've never tried to do something like this... but it would be nice if we could get a `define:vars` for in-lined scripts that still permits some processing.

Not everyone on my team is convinced that custom web elements is a great idea... :sweat_smile: and that tends to be the best way currently to get custom data easily on a mass scale

**Comment by ocavue** [comment]
Hi! Looking forward to seeing official `<script is:inline>` minification support. For those who want this today, I've created an Astro plugin to minify inline scripts. 

https://github.com/ocavue/astro-minify-html-swc

**Comment by jakefb** [comment]
I've created a related feature request https://github.com/withastro/roadmap/discussions/1367

Having the ability to add the `blocking=render` attribute ([mdn link](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/script#blocking)) to processed module scripts (i.e. scripts with `type=module`), would solve many of the use cases requiring inline scripts - i.e. ensuring the script is executed before the page is rendered. The main downside though is it's not currently supported in all browsers.
